File: app-back/Services/DrawService.cs
using CelulaApp.DTOs;
using CelulaApp.Models;
using CelulaApp.Repositories.Interfaces;
using CelulaApp.Services.Interfaces;

namespace CelulaApp.Services;

public class DrawService : IDrawService
{
    private readonly IParticipationRepository _participationRepo;
    private readonly IMemberRepository _memberRepo;
    private readonly IMeetingRepository _meetingRepo;
    private readonly Random _random = new();

    public DrawService(
        IParticipationRepository participationRepo,
        IMemberRepository memberRepo,
        IMeetingRepository meetingRepo)
    {
        _participationRepo = participationRepo;
        _memberRepo = memberRepo;
        _meetingRepo = meetingRepo;
    }

    public async Task<DrawResultDto> DrawMemberAsync(int meetingId)
    {
        var meeting = await _meetingRepo.GetByIdAsync(meetingId)
            ?? throw new KeyNotFoundException($"Encontro {meetingId} não encontrado.");

        var allMembers = await _memberRepo.GetAllAsync();
        if (allMembers.Count == 0)
            throw new InvalidOperationException("Nenhum membro cadastrado.");

        var participations = await _participationRepo.GetByMeetingIdAsync(meetingId);

        // Monta contagem de vezes sorteado por membro neste encontro
        var selectionCount = allMembers.ToDictionary(
            m => m.Id,
            m => participations.FirstOrDefault(p => p.MemberId == m.Id)?.TimesSelected ?? 0
        );

        // Exclui o último sorteado para evitar repetição consecutiva
        // (exceto se só houver um membro)
        var candidates = allMembers.ToList();
        if (meeting.LastDrawnMemberId.HasValue && candidates.Count > 1)
            candidates = candidates.Where(m => m.Id != meeting.LastDrawnMemberId.Value).ToList();

        // Prioriza quem participou menos (menor TimesSelected)
        var minSelections = candidates.Min(m => selectionCount[m.Id]);
        var leastSelected = candidates.Where(m => selectionCount[m.Id] == minSelections).ToList();

        // Sorteio aleatório entre os menos selecionados
        var selected = leastSelected[_random.Next(leastSelected.Count)];

        // Atualiza ou cria participação
        var participation = participations.FirstOrDefault(p => p.MemberId == selected.Id);
        if (participation is null)
        {
            participation = new Participation
            {
                MemberId = selected.Id,
                MeetingId = meetingId,
                TimesSelected = 1,
                TimesSpoken = 0
            };
            await _participationRepo.AddAsync(participation);
        }
        else
        {
            participation.TimesSelected++;
            await _participationRepo.UpdateAsync(participation);
        }

        // Registra último sorteado no encontro
        meeting.LastDrawnMemberId = selected.Id;
        await _meetingRepo.UpdateAsync(meeting);

        return new DrawResultDto(selected.Id, selected.Name, participation.TimesSelected);
    }
}

File: app-backend/src/services/drawService.js
const meetingRepo = require('../repositories/meetingRepository');
const memberRepo = require('../repositories/memberRepository');
const participationRepo = require('../repositories/participationRepository');

async function drawMember(meetingId) {
  const meeting = await meetingRepo.getById(meetingId);
  if (!meeting) throw Object.assign(new Error(`Encontro ${meetingId} não encontrado.`), { status: 404 });

  const allMembers = await memberRepo.getAll();
  if (allMembers.length === 0)
    throw Object.assign(new Error('Nenhum membro cadastrado.'), { status: 400 });

  const participations = await participationRepo.getByMeetingId(meetingId);

  // Contagem de vezes sorteado por membro neste encontro
  const selectionCount = Object.fromEntries(
    allMembers.map((m) => [m.id, participations.find((p) => p.memberId === m.id)?.timesSelected ?? 0])
  );

  // Exclui o último sorteado para evitar repetição consecutiva
  let candidates = [...allMembers];
  if (meeting.lastDrawnMemberId && candidates.length > 1)
    candidates = candidates.filter((m) => m.id !== meeting.lastDrawnMemberId);

  // Prioriza quem foi menos selecionado
  const minSelections = Math.min(...candidates.map((m) => selectionCount[m.id]));
  const leastSelected = candidates.filter((m) => selectionCount[m.id] === minSelections);

  // Sorteio aleatório entre os menos selecionados
  const selected = leastSelected[Math.floor(Math.random() * leastSelected.length)];

  // Atualiza ou cria participação
  const existing = participations.find((p) => p.memberId === selected.id);
  let timesSelected;
  if (!existing) {
    const created = await participationRepo.create({
      memberId: selected.id,
      meetingId,
      timesSelected: 1,
      timesSpoken: 0,
    });
    timesSelected = created.timesSelected;
  } else {
    timesSelected = existing.timesSelected + 1;
    await participationRepo.update(selected.id, meetingId, { timesSelected });
  }

  // Registra último sorteado
  await meetingRepo.update(meetingId, { lastDrawnMemberId: selected.id });

  return { memberId: selected.id, memberName: selected.name, timesSelectedInThisMeeting: timesSelected };
}

module.exports = { drawMember };
